[11113] Pull mode property documentation (#256)
eProsima/Fast-DDS-docs: docs/fastdds/property_policies/non_consolidated_qos.rst

Consequently, this section is prone to frequent updates so the user is advised to check latest changes after upgrading
to a different release version.

.. _push_mode_qos_policy:

> ``Push Mode`` QoS Policy

Reviewer: I'd call the section something like DataWriter operating mode. Also, I think that a link to this section from the datawriter qos section may be appropriate.

Reviewer 2: Done on 21ac05c182abbd86308753aaca6b350f551597c5

^^^^^^^^^^^^^^^^^^^^^^^^

> By default, Fast DDS DataWriters are enabled using the ``push mode``.

Reviewer: ```suggestion
By default, *Fast DDS* |DataWriters| are enabled using ``push mode``.
```

> This implies they will send new samples to matched readers immediately after adding them to their queue.

Reviewer: Their queue here means datawriters', but it could be confused with datareaders'
```suggestion
This implies that they will send new samples to the matched |DataReaders| immediately after adding them to their queue.
```

Reviewer 2: Done on e3a513774fc11e066ef7c8e05e7a6cce7505374d

For writers that produce non periodic bursts of data, this may imply saturating the network with a lot of packets,
increasing the possibility of losing them on unreliable (i.e. UDP) transports.
Depending on their QoS, DataReaders may also have to ignore some received samples, so they will have to be resent.

Configuring the DataWriters on ``pull mode`` offers an alternative by letting each reader pace its own data stream.

> It works by notifying the reader what it is missing, and wait for it to request only as much as it can handle.

Reviewer: ```suggestion
It works by the writer notifying the reader what it is available, and wait for it to request only as much as it can handle.
```

> At the cost of greater latency, this model can deliver reliability while using far fewer packets than in ``push mode``.

Reviewer: ```suggestion
At the cost of greater latency, this model can deliver reliability while using far fewer packets than ``push mode``.
```

> DataWriters periodically announce the state of their queue by means of a heartbeat.

Reviewer: I'd like a reference for the heartbeat

Reviewer 2: Done on b31ded0e91ad900fbf5f775f4651eba187a34ec9

|    :dedent: 8                                                                                                        |
+----------------------------------------------------------------------------------------------------------------------+
| **XML**                                                                                                              |
+----------------------------------------------------------------------------------------------------------------------+
| .. literalinclude:: /../code/XMLTester.xml                                                                           |
|    :language: xml                                                                                                    |
|    :start-after: <!-->PULL_MODE_DATAWRITER<-->                                                                       |
|    :end-before: <!--><-->                                                                                            |
+----------------------------------------------------------------------------------------------------------------------+

> .. note::

Reviewer: I'm missing a note on the behaviour between reliable writer in best effort reader

Reviewer 2: Done on 9f9622f68a00541c25fbf711c8cb4bc757d57c5a

>     Communication to readers running on the same process will always use ``push mode``

Author: ```suggestion
    Communication to readers running on the same process (:ref:`intraprocess-delivery`) will always use ``push mode``
```

I suggest adding the reference in case the user wants to know more.

.. warning::
    It is inconsistent to enable the ``pull mode`` and also set the |ReliabilityQosPolicyKind-api| to
    |BEST_EFFORT_RELIABILITY_QOS-api|.

>     It is inconsistent to enable the ``pull mode`` and also set the |WriterTimes::heartbeatPeriod-api| to

Reviewer: I'd rather have either two warnings or an unordered list here. It looks funny on the HTML

Reviewer 2: Done on 07f2be050dc0fbb63c75b52b847d793a52d53bec

.. _push_mode_qos_policy:

``DataWriter operating mode`` QoS Policy
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, *Fast DDS* |DataWriters| are enabled using ``push mode``.
This implies that they will add new samples into their queue, and then immediately deliver them to matched readers.
For writers that produce non periodic bursts of data, this may imply saturating the network with a lot of packets,
increasing the possibility of losing them on unreliable (i.e. UDP) transports.
Depending on their QoS, DataReaders may also have to ignore some received samples, so they will have to be resent.

Configuring the DataWriters on ``pull mode`` offers an alternative by letting each reader pace its own data stream.

> It works by the writer notifying the reader what it is available, and wait for it to request only as much as it can

Reviewer: ```suggestion
It works by the writer notifying the reader what it is available, and waiting for it to request only as much as it can
```

Reviewer 2: Done in 2535f2ab210ae79568ed2291a522519233bca5df
